### ISSUE TITLE
NumericDate.MarshalJSON overflows on large unix timestamps

### ISSUE THREAD

**Issue by qqiao** [golang-jwt/jwt#199]
Marshalling `NewNumericDate(time.Unix(253402271999, 0))` will result in `-4852145033` instead of `253402271999`

Updates:
This is due to the fact that `Time.UnixNano()` overflows, and `NumericDate.MarshalJSON()` internally uses this method.

PR #200 created to workaround this std library limitation.

**Comment by qqiao** [comment]
This will consistently fail on a M1 Mac:

```
func TestNumericDate_MarshalJSON(t *testing.T) {
	const ts = 253402271999
	expected := fmt.Sprintf("%d", ts)
	date := jwt2.NewNumericDate(time.Unix(253402271999, 0))

	b, err := date.MarshalJSON()
	if err != nil {
		t.Errorf("Failed to marshal numeric date %s. Error: %v",
			date.String(), err)
	}
	got := fmt.Sprintf("%s", b)

	if got != expected {
		t.Errorf("Expected: %s, got: %s", expected, got)
	}
}
```

Output would be `Expected: 253402271999, got: -4852145033`

**Comment by qqiao** [comment]
I figured out the reason: the timestamp is too large, which causes `UnixNano` to overflow. Since MarshalJSON uses UnixNano, it fails too. I'll see if there's a workaround.

**Comment by qqiao** [comment]
PR #200 created, @mfridman could you help review?

**Comment by mfridman** [comment]
Thanks for providing a repro and submitting a PR. 

Indeed that's the result of `UnixNano`, which makes this valid up until 2262. Would using `UnixMicro` be more practical?

**Comment by qqiao** [comment]
I think the PR that I provided would work correctly for all cases, up to the maximum allowed timestamp and would also be future proof.

The issue with using either `nano` or `micro` is that they overflow at some point, which leaves us with cases where a valid `time.Time` cannot be correctly marshalled.

Internally, the `time.Time` implementation stores seconds and nano seconds separately, the PR that I submitted took advantage of that, by formatting them separately and concatinating the strings. That way, any valid `time.Time` would get marshalled correctly.
